{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "df3e9e93",
   "metadata": {},
   "source": [
    "Resources\n",
    "- [xml module documentation](https://docs.python.org/3/library/xml.etree.elementtree.html)\n",
    "- [NCBI APIs documentation](https://www.ncbi.nlm.nih.gov/home/develop/api/)\n",
    "- [MeSH (Medical Subject Headings) Descriptor Search](https://meshb.nlm.nih.gov/search)\n",
    "- [MeSH Documentation](https://pubmed.ncbi.nlm.nih.gov/help/#mhda)\n",
    "- [Pubmed user guide](https://pubmed.ncbi.nlm.nih.gov/help/#mesh-subheadings)\n",
    "\n",
    "Notes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bc96aa3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import xml.etree.ElementTree as ET\n",
    "import webbrowser\n",
    "import pandas as pd\n",
    "import datetime\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "27ddf54d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>annotinfo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>assembly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>biocollections</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>bioproject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>biosample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>biosystems</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>blastdbinfo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>books</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>cdd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>clinvar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>dbvar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>gap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>gapplus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>gds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>geoprofiles</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>grasp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>gtr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>homologene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ipg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>medgen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>mesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ncbisearch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>nlmcatalog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>nuccore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>nucleotide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>omim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>orgtrack</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>pcassay</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>pccompound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>pcsubstance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>pmc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>popset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>proteinclusters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>protfam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>pubmed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>seqannot</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>sra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>structure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>taxonomy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                dbs\n",
       "0         annotinfo\n",
       "1          assembly\n",
       "2    biocollections\n",
       "3        bioproject\n",
       "4         biosample\n",
       "5        biosystems\n",
       "6       blastdbinfo\n",
       "7             books\n",
       "8               cdd\n",
       "9           clinvar\n",
       "10            dbvar\n",
       "11              gap\n",
       "12          gapplus\n",
       "13              gds\n",
       "14             gene\n",
       "15           genome\n",
       "16      geoprofiles\n",
       "17            grasp\n",
       "18              gtr\n",
       "19       homologene\n",
       "20              ipg\n",
       "21           medgen\n",
       "22             mesh\n",
       "23       ncbisearch\n",
       "24       nlmcatalog\n",
       "25          nuccore\n",
       "26       nucleotide\n",
       "27             omim\n",
       "28         orgtrack\n",
       "29          pcassay\n",
       "30       pccompound\n",
       "31      pcsubstance\n",
       "32              pmc\n",
       "33           popset\n",
       "34          protein\n",
       "35  proteinclusters\n",
       "36          protfam\n",
       "37           pubmed\n",
       "38         seqannot\n",
       "39              snp\n",
       "40              sra\n",
       "41        structure\n",
       "42         taxonomy"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list of databases accessible to Entrez\n",
    "r = requests.get('https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi')\n",
    "dom = ET.fromstring(r.text)\n",
    "dbs = dom.find('DbList')\n",
    "dbs_names = []\n",
    "for db in dbs.findall('DbName'):\n",
    "    dbs_names.append(db.text)\n",
    "dbs_names.sort()\n",
    "pd.DataFrame(dbs_names, columns=['dbs'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7fe6d45",
   "metadata": {},
   "source": [
    "### Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "293979e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_article_index(articles, PMID):\n",
    "    # Find an index of specific article in our articles based on PMID\n",
    "    # Params:\n",
    "    #    articles: array of articles with type xml.etree.ElementTree.Element\n",
    "    #    PMID: string\n",
    "    # Example:\n",
    "    #    idx = get_article_index(articles, '34163160')\n",
    "    for i in range(len(articles)):\n",
    "        if PMID == articles[i].find('MedlineCitation/PMID').text:\n",
    "            return i\n",
    "    return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aed7e3c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_article_biomarker(biomarker):\n",
    "    # Get the title, keyword list, chemical list, and abstract of all the articles, \n",
    "    # related to the biomarkers specified related with MDD and return a dataframe\n",
    "    #\n",
    "    # Params:\n",
    "    #     biomarker: the biomarker to be searched for\n",
    "    # Variables:\n",
    "    #    ...\n",
    "    \n",
    "    # esearch for matching articles\n",
    "    payload = {'db': 'pubmed',\n",
    "           'term': 'Major depressive disorder[MH] ' + biomarker,\n",
    "           'datetype':'edat',\n",
    "           'usehistory': 'y'}\n",
    "    r = requests.get('https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi', params=payload)\n",
    "    dom = ET.fromstring(r.text)\n",
    "    web_env = dom.find('WebEnv').text\n",
    "    query_key = dom.find('QueryKey').text\n",
    "    assert web_env != None\n",
    "    assert query_key != None\n",
    "    \n",
    "    # efetch to get details about the article\n",
    "    payload = {'db': 'pubmed',\n",
    "           'query_key': query_key,\n",
    "           'WebEnv': web_env,\n",
    "           'rettype': 'xml',\n",
    "           'retstart': 0}\n",
    "    r = requests.get('https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi', params=payload)\n",
    "    dom = ET.fromstring(r.text)\n",
    "    articles = dom.findall('PubmedArticle')\n",
    "    \n",
    "    # initialize df\n",
    "    df = pd.DataFrame(columns=['title', 'pmid', 'language', 'date', 'mesh major', 'keyword', 'mesh minor',\n",
    "                              'mesh qualifier', 'chemical', 'background', 'objective',\n",
    "                              'method', 'result', 'discussion', 'conclusion'])\n",
    "    # begin parsing the XML of each article\n",
    "    for idx in range(len(articles)):\n",
    "        article = articles[idx].find('MedlineCitation')\n",
    "        abstract = article.find('Article/Abstract')\n",
    "        title = article.find('Article/ArticleTitle')\n",
    "        # concatenating broken texts due to tags\n",
    "        title_tail = [(x.text or \"\") + (x.tail or \"\") for x in title.findall('.//')]\n",
    "        title_tail = ''.join(title_tail)\n",
    "        title = (title.text or \"\") + title_tail\n",
    "        PMID = article.find('PMID').text\n",
    "        date = article.find('DateRevised')\n",
    "        date = datetime.datetime(int(date.find('Year').text), int(date.find('Month').text), int(date.find('Day').text))\n",
    "        language = article.find('Article/Language').text\n",
    "        # minor relevant MeSH terms\n",
    "        meshMinorList = []\n",
    "        # major relevant MeSH terms\n",
    "        meshMajorList = []\n",
    "        # qualifer relevant MeSH terms\n",
    "        meshQualifierList = set()\n",
    "        keyWordList = []\n",
    "        chemicalList = []\n",
    "        if article.find('MeshHeadingList'):\n",
    "            for meshHeading in article.find('MeshHeadingList'):\n",
    "                for item in meshHeading:\n",
    "                    if item.tag == 'QualifierName':\n",
    "                        meshQualifierList.add(item.text)\n",
    "                    else:      \n",
    "                        if 'Y' == item.attrib['MajorTopicYN']:\n",
    "                            meshMajorList.append(item.text)\n",
    "                        else:\n",
    "                            meshMinorList.append(item.text)\n",
    "        if article.find('KeywordList'):\n",
    "            for keyword in article.find('KeywordList'):\n",
    "                keyWordList.append(keyword.text)\n",
    "        if article.find('ChemicalList'):\n",
    "            for chemical in article.find('ChemicalList'):\n",
    "                for item in chemical:\n",
    "                    if item.tag == 'NameOfSubstance':\n",
    "                        chemicalList.append(item.text)\n",
    "        meshQualifierList = list(meshQualifierList)\n",
    "        \n",
    "        # for article that has abstract\n",
    "        if abstract != None:\n",
    "            background, objective, method, result, discussion, conclusion = '', '', '', '', '', ''\n",
    "            abstract_sections = abstract.findall('AbstractText')\n",
    "            for section in abstract_sections:\n",
    "                sub_sections_text = [(x.text or \"\") + (x.tail or \"\") for x in section.findall('.//')]\n",
    "                sub_sections_text = ''.join(sub_sections_text)\n",
    "                abstract_text = (section.text or \"\") + sub_sections_text\n",
    "                if 'NlmCategory' in section.attrib or 'Label' in section.attrib:\n",
    "                    attrib = section.attrib\n",
    "                    temp = attrib.get('NlmCategory', attrib['Label'])\n",
    "                    if temp == 'UNASSIGNED':\n",
    "                        temp = attrib['Label']\n",
    "                    attrib = temp\n",
    "                    attrib = attrib.lower().strip()\n",
    "                    if 'background' in attrib:\n",
    "                        background = abstract_text\n",
    "                    elif 'objective' in attrib or 'aim' in attrib:\n",
    "                        objective = abstract_text\n",
    "                    elif 'method' in attrib:\n",
    "                        method = abstract_text\n",
    "                    elif 'result' in attrib:\n",
    "                        result = abstract_text\n",
    "                    elif 'discussion' in attrib:\n",
    "                        discussion = abstract_text\n",
    "                    elif 'conclusion' in attrib:\n",
    "                        conclusion = abstract_text\n",
    "                # the abstract section doesn't have a label and that's the only abstract section\n",
    "                elif len(abstract_sections) == 1:\n",
    "                    # parse articles that use <b> tag for abstract sections\n",
    "                    section_tags = list(section)\n",
    "                    section_tag_names = [x.tag for x in section_tags]\n",
    "                    if 'b' in section_tag_names:\n",
    "                        count = 0\n",
    "                        while count < len(section_tags):\n",
    "                            abstract_text = \"\"\n",
    "                            if section_tag_names[count] == 'b':\n",
    "                                attrib = section_tags[count].text\n",
    "                                # if the attrib has an inner tag which has the text\n",
    "                                if attrib is None:\n",
    "                                    attrib = \"\".join([x.text for x in section_tags[count]])\n",
    "                                attrib = attrib.lower().strip()\n",
    "                                abstract_text += section_tags[count].tail or \"\"\n",
    "                                if count < len(section_tags) - 1:\n",
    "                                    count += 1\n",
    "                                    while count < len(section_tags) and section_tag_names[count] != 'b':\n",
    "                                        abstract_text += section_tags[count].text or \"\"\n",
    "                                        abstract_text += section_tags[count].tail or \"\"\n",
    "                                        count += 1\n",
    "                                    count -= 1\n",
    "                                if 'background' in attrib:\n",
    "                                    background = abstract_text\n",
    "                                elif 'objective' in attrib or 'aim' in attrib:\n",
    "                                    objective = abstract_text\n",
    "                                elif 'method' in attrib:\n",
    "                                    method = abstract_text\n",
    "                                elif 'result' in attrib:\n",
    "                                    result = abstract_text\n",
    "                                elif 'discussion' in attrib:\n",
    "                                    discussion = abstract_text\n",
    "                                elif 'conclusion' in attrib:\n",
    "                                    conclusion = abstract_text\n",
    "                            count += 1\n",
    "\n",
    "                    else:\n",
    "                        background = abstract_text\n",
    "            df = df.append({'title': title, 'pmid': PMID, 'language': language, 'date': date, \n",
    "                   'mesh major': meshMajorList, 'keyword': keyWordList, 'mesh minor': meshMinorList, 'mesh qualifier': meshQualifierList, 'chemical': chemicalList, \n",
    "                   'background': background, 'objective': objective, 'method': method, 'result': result, \n",
    "                   'discussion': discussion, 'conclusion': conclusion}, ignore_index=True)\n",
    "        else:\n",
    "            pass\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "03627924",
   "metadata": {},
   "outputs": [],
   "source": [
    "def display_abstract(background='', objective='', method='', result='', discussion='', conclusion=''):\n",
    "    # Print out all fields of abstract that are not empty\n",
    "    # Params: string\n",
    "    print('[Background]')\n",
    "    if len(background) > 0:\n",
    "        print(background)\n",
    "    print('[Objective]')\n",
    "    if len(objective) > 0:\n",
    "        print(objective)\n",
    "    print('[Method]')\n",
    "    if len(method) > 0:\n",
    "        print(method)\n",
    "    print('[Result]')\n",
    "    if len(result) > 0:\n",
    "        print(result)\n",
    "    print('[Discussion]')\n",
    "    if len(discussion) > 0:\n",
    "        print(discussion)\n",
    "    print('[Conclusion]')\n",
    "    if len(conclusion) > 0:\n",
    "        print(conclusion)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2196ff4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def browse_articles(articles, idx=0):\n",
    "    # Get the title, keyword list, chemical list, and abstract of one of the articles, rerun this cell to iterate through the articles\n",
    "    # Params:\n",
    "    #     idx: starting index to browese the articles, default to the beginning (i.e. 0)\n",
    "    #     articles: array of articles with type xml.etree.ElementTree.Element\n",
    "    article = articles[idx].find('MedlineCitation')\n",
    "    abstract = article.find('Article/Abstract')\n",
    "    title = article.find('Article/ArticleTitle')\n",
    "    # concatenating broken texts due to tags\n",
    "    title_tail = [(x.text or \"\") + (x.tail or \"\") for x in title.findall('.//')]\n",
    "    title_tail = ''.join(title_tail)\n",
    "    title = (title.text or \"\") + title_tail\n",
    "    PMID = article.find('PMID').text\n",
    "    date = article.find('DateRevised')\n",
    "    date = datetime.datetime(int(date.find('Year').text), int(date.find('Month').text), int(date.find('Day').text))\n",
    "    language = article.find('Article/Language').text\n",
    "    # minor relevant MeSH terms\n",
    "    meshMinorList = []\n",
    "    # major relevant MeSH terms\n",
    "    meshMajorList = []\n",
    "    # qualifer relevant MeSH terms\n",
    "    meshQualifierList = set()\n",
    "    keyWordList = []\n",
    "    chemicalList = []\n",
    "    if article.find('MeshHeadingList'):\n",
    "        for meshHeading in article.find('MeshHeadingList'):\n",
    "            for item in meshHeading:\n",
    "                if item.tag == 'QualifierName':\n",
    "                    meshQualifierList.add(item.text)\n",
    "                else:      \n",
    "                    if 'Y' == item.attrib['MajorTopicYN']:\n",
    "                        meshMajorList.append(item.text)\n",
    "                    else:\n",
    "                        meshMinorList.append(item.text)\n",
    "    if article.find('KeywordList'):\n",
    "        for keyword in article.find('KeywordList'):\n",
    "            keyWordList.append(keyword.text)\n",
    "    if article.find('ChemicalList'):\n",
    "        for chemical in article.find('ChemicalList'):\n",
    "            for item in chemical:\n",
    "                if item.tag == 'NameOfSubstance':\n",
    "                    chemicalList.append(item.text)\n",
    "\n",
    "    print('Title:', title)\n",
    "    print('PMID:', PMID)\n",
    "    print(\"Language:\", language)\n",
    "    print(date.strftime(\"Date Revised: %m/%d/%Y\"))\n",
    "    print(\"Keywords:\", keyWordList)\n",
    "    print(\"MeSH Major:\", meshMajorList)\n",
    "    print(\"MeSH Minor:\", meshMinorList)\n",
    "    print(\"MeSH Qualifier:\", meshQualifierList)\n",
    "    print(\"Chemicals:\", chemicalList)\n",
    "    print('='*50 + 'Abstract' + '='*50)\n",
    "    \n",
    "    # for article that has abstract\n",
    "    if abstract != None:\n",
    "        background, objective, method, result, discussion, conclusion = '', '', '', '', '', ''\n",
    "        abstract_sections = abstract.findall('AbstractText')\n",
    "        for section in abstract_sections:\n",
    "            sub_sections_text = [(x.text or \"\") + (x.tail or \"\") for x in section.findall('.//')]\n",
    "            sub_sections_text = ''.join(sub_sections_text)\n",
    "            abstract_text = (section.text or \"\") + sub_sections_text\n",
    "            if 'NlmCategory' in section.attrib or 'Label' in section.attrib:\n",
    "                attrib = section.attrib\n",
    "                temp = attrib.get('NlmCategory', attrib['Label'])\n",
    "                if temp == 'UNASSIGNED':\n",
    "                    temp = attrib['Label']\n",
    "                attrib = temp\n",
    "                attrib = attrib.lower().strip()\n",
    "                if 'background' in attrib:\n",
    "                    background = abstract_text\n",
    "                elif 'objective' in attrib or 'aim' in attrib:\n",
    "                    objective = abstract_text\n",
    "                elif 'method' in attrib:\n",
    "                    method = abstract_text\n",
    "                elif 'result' in attrib:\n",
    "                    result = abstract_text\n",
    "                elif 'discussion' in attrib:\n",
    "                    discussion = abstract_text\n",
    "                elif 'conclusion' in attrib:\n",
    "                    conclusion = abstract_text\n",
    "            # the abstract section doesn't have a label and that's the only abstract section\n",
    "            elif len(abstract_sections) == 1:\n",
    "                # parse articles that use <b> tag for abstract sections\n",
    "                section_tags = list(section)\n",
    "                section_tag_names = [x.tag for x in section_tags]\n",
    "                if 'b' in section_tag_names:\n",
    "                    count = 0\n",
    "                    while count < len(section_tags):\n",
    "                        abstract_text = \"\"\n",
    "                        if section_tag_names[count] == 'b':\n",
    "                            attrib = section_tags[count].text\n",
    "                            # if the attrib has an inner tag which has the text\n",
    "                            if attrib is None:\n",
    "                                attrib = \"\".join([x.text for x in section_tags[count]])\n",
    "                            attrib = attrib.lower().strip()\n",
    "                            abstract_text += section_tags[count].tail or \"\"\n",
    "                            if count < len(section_tags) - 1:\n",
    "                                count += 1\n",
    "                                while count < len(section_tags) and section_tag_names[count] != 'b':\n",
    "                                    abstract_text += section_tags[count].text or \"\"\n",
    "                                    abstract_text += section_tags[count].tail or \"\"\n",
    "                                    count += 1\n",
    "                                count -= 1\n",
    "                            if 'background' in attrib:\n",
    "                                background = abstract_text\n",
    "                            elif 'objective' in attrib or 'aim' in attrib:\n",
    "                                objective = abstract_text\n",
    "                            elif 'method' in attrib:\n",
    "                                method = abstract_text\n",
    "                            elif 'result' in attrib:\n",
    "                                result = abstract_text\n",
    "                            elif 'discussion' in attrib:\n",
    "                                discussion = abstract_text\n",
    "                            elif 'conclusion' in attrib:\n",
    "                                conclusion = abstract_text\n",
    "                        count += 1\n",
    "\n",
    "                else:\n",
    "                    background = abstract_text\n",
    "        display_abstract(background=background, objective=objective, method=method, result=result, discussion=discussion, conclusion=conclusion)\n",
    "        print()\n",
    "    else:\n",
    "        print(\"No abstract available for this article\")\n",
    "    idx += 1\n",
    "    idx %= len(articles)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddf267c8",
   "metadata": {},
   "source": [
    "### Cancer Study using MeSH"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85fbc3df",
   "metadata": {},
   "source": [
    "Get the number of articles related to cancer on pubmed in the past 3 days"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "acd7e9dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "payload = {'db': 'pubmed',\n",
    "           'term': 'cancer[MAJR]',\n",
    "           'datetype':'edat',\n",
    "           'usehistory': 'y',\n",
    "           'reldate': 3}\n",
    "r = requests.get('https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi', params=payload)\n",
    "dom = ET.fromstring(r.text)\n",
    "web_env = dom.find('WebEnv').text\n",
    "query_key = dom.find('QueryKey').text\n",
    "assert web_env != None\n",
    "assert query_key != None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "65116100",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of results on pubmed related to cancer: 43\n"
     ]
    }
   ],
   "source": [
    "print('# of results on pubmed related to cancer:', dom.find('Count').text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4eb304e4",
   "metadata": {},
   "source": [
    "Get the metadata summary of the last result on pubmed regarding cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "95c9aa18",
   "metadata": {},
   "outputs": [],
   "source": [
    "payload = {'query_key': query_key,\n",
    "           'WebEnv': web_env,\n",
    "           'version': 2.0} \n",
    "r = requests.get('https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi', params=payload)\n",
    "dom = ET.fromstring(r.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1094de12",
   "metadata": {},
   "source": [
    "Get the article summary on pubmed regarding cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "41b2a1a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "payload = {'db': 'pubmed',\n",
    "           'query_key': query_key,\n",
    "           'WebEnv': web_env,\n",
    "           'rettype': 'xml',\n",
    "           'retstart': 0}\n",
    "r = requests.get('https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi', params=payload)\n",
    "dom = ET.fromstring(r.text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "68496750",
   "metadata": {},
   "outputs": [],
   "source": [
    "# idx to update the article for the next cell\n",
    "idx = 0\n",
    "articles = dom.findall('PubmedArticle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d4e3250c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Title: Laparoscopic lateral lymph node dissection in two fascial spaces for locally advanced lower rectal cancer.\n",
      "PMID: 34239276\n",
      "Language: eng\n",
      "Date Revised: 07/12/2021\n",
      "Keywords: ['Cardinal ligament', 'Fascial anatomy', 'Lateral lymph node dissection', 'Locally advanced low rectal cancer', 'Vesicohypogastric fascia', 'Visceral fascia']\n",
      "MeSH Major: ['Laparoscopy', 'Rectal Neoplasms']\n",
      "MeSH Minor: ['Dissection', 'Humans', 'Lymph Node Excision', 'Lymph Nodes']\n",
      "MeSH Qualifier: {'adverse effects', 'surgery'}\n",
      "Chemicals: []\n",
      "==================================================Abstract==================================================\n",
      "[Background]\n",
      "The procedure for lateral lymph node (LLN) dissection (LLND) is complicated and can result in complications. We developed a technique for laparoscopic LLND based on two fascial spaces to simplify the procedure.\n",
      "[Objective]\n",
      "To clarify the anatomical basis of laparoscopic LLND in two fascial spaces and to evaluate its efficacy and safety in treating locally advanced low rectal cancer (LALRC).\n",
      "[Method]\n",
      "Cadaveric dissection was performed on 24 pelvises, and the fascial composition related to LLND was observed and described. Three dimensional-laparoscopic total mesorectal excision with LLND was performed in 20 patients with LALRC, and their clinical data were analyzed.\n",
      "[Result]\n",
      "The cadaver study showed that the fascia propria of the rectum, urogenital fascia, vesicohypogastric fascia and parietal fascia lie side by side in a medial-lateral direction constituting the dissection plane for curative rectal cancer surgery, and the last three fasciae formed two spaces (Latzko's pararectal space and paravesical space) which were the surgical area for LLND. Laparoscopic LLND in two fascial spaces was performed successfully in all 20 patients. The median operating time, blood loss and postoperative hospitalization were 178 (152-243) min, 55 (25-150) mL and 10 (7-20) d, respectively. The median number of harvested LLNs was 8.6 (6-12), and pathologically positive LLN metastasis was confirmed in 7 (35.0%) cases. Postoperative complications included lower limb pain in 1 case and lymph leakage in 1 case.\n",
      "[Discussion]\n",
      "[Conclusion]\n",
      "Our preliminary surgical experience suggests that laparoscopic LLND based on fascial spaces is a feasible, effective and safe procedure for treating LALRC.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "browse_articles(articles, idx=idx)\n",
    "idx += 1\n",
    "idx %= len(articles)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "922d75f4",
   "metadata": {},
   "source": [
    "Uncommend cell below to view the hierarchy of the xml file in a new tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "24eddcd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# webbrowser.open(r.url);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f64f224",
   "metadata": {},
   "source": [
    "### Depression & Biomarker Study using MeSH"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff7b89e8",
   "metadata": {},
   "source": [
    "Get the number of articles regarding major depressive disorder and leptin on pubmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f9542c9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "payload = {'db': 'pubmed',\n",
    "           'term': 'Major depressive disorder[MH] leptin',\n",
    "           'datetype':'edat',\n",
    "           'usehistory': 'y'}\n",
    "r = requests.get('https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi', params=payload)\n",
    "dom = ET.fromstring(r.text)\n",
    "web_env = dom.find('WebEnv').text\n",
    "query_key = dom.find('QueryKey').text\n",
    "assert web_env != None\n",
    "assert query_key != None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7747fb9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of results on pubmed related to depression and biomarker: 117\n"
     ]
    }
   ],
   "source": [
    "print('# of results on pubmed related to depression and biomarker:', dom.find('Count').text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "114d2071",
   "metadata": {},
   "source": [
    "Get the article summary "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d1aa4a07",
   "metadata": {},
   "outputs": [],
   "source": [
    "payload = {'db': 'pubmed',\n",
    "           'query_key': query_key,\n",
    "           'WebEnv': web_env,\n",
    "           'rettype': 'xml',\n",
    "           'retstart': 0}\n",
    "r = requests.get('https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi', params=payload)\n",
    "dom = ET.fromstring(r.text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7222a04b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# idx to update the article for the next cell\n",
    "idx = 0\n",
    "articles = dom.findall('PubmedArticle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "50800618",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Title: Plasma leptin correlates with anthranilic acid in schizophrenia but not in major depressive disorder.\n",
      "PMID: 32855025\n",
      "Language: eng\n",
      "Date Revised: 01/22/2021\n",
      "Keywords: []\n",
      "MeSH Major: ['Depressive Disorder, Major', 'Schizophrenia']\n",
      "MeSH Minor: ['Humans', 'Insulin', 'Leptin', 'Plasma', 'Proteomics', 'ortho-Aminobenzoates']\n",
      "MeSH Qualifier: {'drug therapy'}\n",
      "Chemicals: ['Insulin', 'Leptin', 'ortho-Aminobenzoates', 'anthranilic acid']\n",
      "==================================================Abstract==================================================\n",
      "No abstract available for this article\n"
     ]
    }
   ],
   "source": [
    "browse_articles(articles, idx=idx)\n",
    "idx += 1\n",
    "idx %= len(articles)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c61cafdd",
   "metadata": {},
   "source": [
    "### Begin scraping pubmed articles in regards to different candidataes of biomarkers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9cea9dda",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BDNF\n",
      "IL-1\n",
      "IL-6\n",
      "TNF\n",
      "Malondialdehyde\n",
      "Superoxide dismutase\n",
      "corticosteroids\n",
      "leptin\n"
     ]
    }
   ],
   "source": [
    "biomarkers = ['BDNF', 'IL-1', 'IL-6', 'TNF', 'Malondialdehyde', 'Superoxide dismutase', 'corticosteroids', 'leptin']\n",
    "df_dict = {}\n",
    "for biomarker in biomarkers:\n",
    "    print(biomarker)\n",
    "    df_dict[biomarker] = get_article_biomarker(biomarker)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "344df0e3",
   "metadata": {},
   "source": [
    "Get the number of articles of each biomarkder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7c61cae5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BDNF: 1104\n",
      "IL-1: 270\n",
      "IL-6: 651\n",
      "TNF: 402\n",
      "Malondialdehyde: 65\n",
      "Superoxide dismutase: 79\n",
      "corticosteroids: 3098\n",
      "leptin: 109\n"
     ]
    }
   ],
   "source": [
    "for key in df_dict.keys():\n",
    "    print(f\"{key}: {df_dict[key].shape[0]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b42c7c7c",
   "metadata": {},
   "source": [
    "Take a look at cortisols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fcb31f84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>pmid</th>\n",
       "      <th>language</th>\n",
       "      <th>date</th>\n",
       "      <th>mesh major</th>\n",
       "      <th>keyword</th>\n",
       "      <th>mesh minor</th>\n",
       "      <th>mesh qualifier</th>\n",
       "      <th>chemical</th>\n",
       "      <th>background</th>\n",
       "      <th>objective</th>\n",
       "      <th>method</th>\n",
       "      <th>result</th>\n",
       "      <th>discussion</th>\n",
       "      <th>conclusion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>The Role of HPA Axis and Allopregnanolone on t...</td>\n",
       "      <td>34071053</td>\n",
       "      <td>eng</td>\n",
       "      <td>2021-06-21</td>\n",
       "      <td>[]</td>\n",
       "      <td>[PTSD, brexanolone, depression, hypothalamus-p...</td>\n",
       "      <td>[Adaptation, Physiological, Animals, Antidepre...</td>\n",
       "      <td>[psychology, physiopathology, drug therapy, me...</td>\n",
       "      <td>[Antidepressive Agents, GABA-A Receptor Agonis...</td>\n",
       "      <td>Under stressful conditions, the hypothalamic-p...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TNIP2 mediates GRβ-promoted inflammation and i...</td>\n",
       "      <td>33932528</td>\n",
       "      <td>eng</td>\n",
       "      <td>2021-06-23</td>\n",
       "      <td>[Depressive Disorder, Major, Receptors, Glucoc...</td>\n",
       "      <td>[Glucocorticoid receptor β, Major depressive d...</td>\n",
       "      <td>[Adaptor Proteins, Signal Transducing, Glucoco...</td>\n",
       "      <td>[metabolism]</td>\n",
       "      <td>[Adaptor Proteins, Signal Transducing, Glucoco...</td>\n",
       "      <td>In depression, continual activation of the hyp...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>The effects of Pythagorean Self-Awareness Inte...</td>\n",
       "      <td>33894540</td>\n",
       "      <td>eng</td>\n",
       "      <td>2021-07-05</td>\n",
       "      <td>[Depressive Disorder, Major]</td>\n",
       "      <td>[Cortisol, Depression, Major depressive disord...</td>\n",
       "      <td>[Adult, Anxiety, Depression, Humans, Hydrocort...</td>\n",
       "      <td>[therapy]</td>\n",
       "      <td>[Hydrocortisone]</td>\n",
       "      <td>Stress plays an important role in major depres...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Potassium-Titanyl-Phosphate (KTP) Laser Photoc...</td>\n",
       "      <td>33753714</td>\n",
       "      <td>eng</td>\n",
       "      <td>2021-05-14</td>\n",
       "      <td>[Depressive Disorder, Major, Hemangioma, Laser...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Humans, Light Coagulation, Male, Middle Aged,...</td>\n",
       "      <td>[therapeutic use, surgery]</td>\n",
       "      <td>[Phosphates, Potassium]</td>\n",
       "      <td>BACKGROUND Hemangiomas are relatively rare, sl...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Role of insulin-like growth factor 1, sex and ...</td>\n",
       "      <td>33731067</td>\n",
       "      <td>eng</td>\n",
       "      <td>2021-04-26</td>\n",
       "      <td>[Depressive Disorder, Major]</td>\n",
       "      <td>[Cortisol, DHEAS, Estradiol, IGF1, Testosterone]</td>\n",
       "      <td>[Dehydroepiandrosterone Sulfate, Humans, Hydro...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Testosterone, Dehydroepiandrosterone Sulfate,...</td>\n",
       "      <td>Hormones of the hypothalamic-pituitary-gonadal...</td>\n",
       "      <td></td>\n",
       "      <td>Serum estradiol, testosterone, cortisol, DHEAS...</td>\n",
       "      <td>Patients had significantly lower estradiol lev...</td>\n",
       "      <td></td>\n",
       "      <td>Estradiol may affect the pathogenesis and seve...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               title      pmid language  \\\n",
       "0  The Role of HPA Axis and Allopregnanolone on t...  34071053      eng   \n",
       "1  TNIP2 mediates GRβ-promoted inflammation and i...  33932528      eng   \n",
       "2  The effects of Pythagorean Self-Awareness Inte...  33894540      eng   \n",
       "3  Potassium-Titanyl-Phosphate (KTP) Laser Photoc...  33753714      eng   \n",
       "4  Role of insulin-like growth factor 1, sex and ...  33731067      eng   \n",
       "\n",
       "        date                                         mesh major  \\\n",
       "0 2021-06-21                                                 []   \n",
       "1 2021-06-23  [Depressive Disorder, Major, Receptors, Glucoc...   \n",
       "2 2021-07-05                       [Depressive Disorder, Major]   \n",
       "3 2021-05-14  [Depressive Disorder, Major, Hemangioma, Laser...   \n",
       "4 2021-04-26                       [Depressive Disorder, Major]   \n",
       "\n",
       "                                             keyword  \\\n",
       "0  [PTSD, brexanolone, depression, hypothalamus-p...   \n",
       "1  [Glucocorticoid receptor β, Major depressive d...   \n",
       "2  [Cortisol, Depression, Major depressive disord...   \n",
       "3                                                 []   \n",
       "4   [Cortisol, DHEAS, Estradiol, IGF1, Testosterone]   \n",
       "\n",
       "                                          mesh minor  \\\n",
       "0  [Adaptation, Physiological, Animals, Antidepre...   \n",
       "1  [Adaptor Proteins, Signal Transducing, Glucoco...   \n",
       "2  [Adult, Anxiety, Depression, Humans, Hydrocort...   \n",
       "3  [Humans, Light Coagulation, Male, Middle Aged,...   \n",
       "4  [Dehydroepiandrosterone Sulfate, Humans, Hydro...   \n",
       "\n",
       "                                      mesh qualifier  \\\n",
       "0  [psychology, physiopathology, drug therapy, me...   \n",
       "1                                       [metabolism]   \n",
       "2                                          [therapy]   \n",
       "3                         [therapeutic use, surgery]   \n",
       "4                                                 []   \n",
       "\n",
       "                                            chemical  \\\n",
       "0  [Antidepressive Agents, GABA-A Receptor Agonis...   \n",
       "1  [Adaptor Proteins, Signal Transducing, Glucoco...   \n",
       "2                                   [Hydrocortisone]   \n",
       "3                            [Phosphates, Potassium]   \n",
       "4  [Testosterone, Dehydroepiandrosterone Sulfate,...   \n",
       "\n",
       "                                          background objective  \\\n",
       "0  Under stressful conditions, the hypothalamic-p...             \n",
       "1  In depression, continual activation of the hyp...             \n",
       "2  Stress plays an important role in major depres...             \n",
       "3  BACKGROUND Hemangiomas are relatively rare, sl...             \n",
       "4  Hormones of the hypothalamic-pituitary-gonadal...             \n",
       "\n",
       "                                              method  \\\n",
       "0                                                      \n",
       "1                                                      \n",
       "2                                                      \n",
       "3                                                      \n",
       "4  Serum estradiol, testosterone, cortisol, DHEAS...   \n",
       "\n",
       "                                              result discussion  \\\n",
       "0                                                                 \n",
       "1                                                                 \n",
       "2                                                                 \n",
       "3                                                                 \n",
       "4  Patients had significantly lower estradiol lev...              \n",
       "\n",
       "                                          conclusion  \n",
       "0                                                     \n",
       "1                                                     \n",
       "2                                                     \n",
       "3                                                     \n",
       "4  Estradiol may affect the pathogenesis and seve...  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dict[\"corticosteroids\"].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2124d7cd",
   "metadata": {},
   "source": [
    "### Export the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9930b09c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['.ipynb_checkpoints', 'data', 'Data mining for papers.ipynb']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "02a6510a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"data\" not in os.listdir():\n",
    "    os.mkdir(\"data\")\n",
    "for key in df_dict.keys():\n",
    "    if os.path.exists()\n",
    "    df_dict[key].to_pickle(f\"data/{key}.pkl\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
